src/database-controller/src/core/k8s.js

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+const k8s = require('@kubernetes/client-node');
+const logger = require('./logger');
+const config = require('./config');
+const { timeoutDecorator } = require('./util');
+const kc = new k8s.KubeConfig();
+
+if (config.rbacEnabled) {
+  // If RBAC is enabled, we can use kc.loadFromDefault() to load k8s config in containers.
+  // For local debugging purpose, one can set CUSTOM_K8S_API_SERVER_URL, CUSTOM_K8S_CA_FILE and CUSTOM_K8S_TOKEN_FILE,
+  // to connect to RBAC k8s cluster. Ca and Token file should be from a valid service account.
+  if (config.customK8sApiServerURL) {
+    const cluster = {
+      name: 'inCluster',
+      caFile: config.customK8sCaFile,
+      server: config.customK8sApiServerURL,
+      skipTLSVerify: false,
+    };
+    const user = {
+      name: 'inClusterUser',
+      authProvider: {
+        name: 'tokenFile',
+        config: {
+          tokenFile: config.customK8sTokenFile,
+        },
+      },
+    };
+    kc.loadFromClusterAndUser(cluster, user);
+  } else {
+    kc.loadFromDefault();
+  }
+} else {
+  // If RBAC is not enabled, use CUSTOM_K8S_API_SERVER_URL to connect to API server.
+  const cluster = { name: 'cluster', server: config.customK8sApiServerURL };
+  const user = { name: 'user' };
+  kc.loadFromClusterAndUser(cluster, user);
+}
+
+// If API server reports a non-200 status code, the client will
+// throw an error. In such case, error.name will be "HttpError",
+// and error.response.statusCode is the actual status code.
+// If network is disconnected, the error will be a different one,
+// and you cannot get error.name and error.statusCode.
+const customObjectsClient = kc.makeApiClient(k8s.CustomObjectsApi);
+
+async function getFramework(name, namespace = 'default') {
+  const res = await customObjectsClient.getNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+    name,
+  );
+  return res.response;
+}
+
+async function listFramework(name, namespace = 'default') {
+  const res = await customObjectsClient.listNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+  );
+  return res.response;
+}
+
+async function createFramework(frameworkDescription, namespace = 'default') {
+  const res = await customObjectsClient.createNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+    frameworkDescription,
+  );
+  return res.response;
+}
+
+async function patchFramework(name, data, namespace = 'default') {
+  const res = await customObjectsClient.patchNamespacedCustomObject(
+    'frameworkcontroller.microsoft.com',
+    'v1',
+    namespace,
+    'frameworks',
+    name,
+    data,
```

Comment:
>data [](start = 4, length = 4)

Make sure the data does not contain status field by explicitly set data.status = null

Reply:
Done, using `delete data.status`